### PR TITLE
GROK-16001: EDA: Softmax

### PR DESCRIPTION
**Pull request by vmakarichev**
Added softmax classifier

### CODE REVIEW

File: packages/EDA/src/package.ts [hunk 1]
```diff
@@ -542,3 +546,213 @@ export function anova(table: DG.DataFrame, factor: DG.Column, feature: DG.Column
...
+//description: Linear regression demo
+//input: dataframe table
+//input: column_list features {type: numerical}
+//input: column target {type: numerical}
+//input: bool plot = true {caption: plot}
+export async function linearRegression(table: DG.DataFrame, features: DG.ColumnList, target: DG.Column, plot: boolean): Promise<void> {
+  const t1 = performance.now();
+  const params = await getLinearRegressionParams(features, target);
+  const t2 = performance.now();
+  console.log(`Fit: ${t2 - t1} ms.`);
```
Comment (skalkin) [inline packages/EDA/src/package.ts:560]: Let's not write to console in the production code

File: packages/EDA/src/package.ts [hunk 2]
```diff
@@ -542,3 +546,213 @@ export function anova(table: DG.DataFrame, factor: DG.Column, feature: DG.Column
 export function kNNImputation() {
   runKNNImputer();
 }
+
+//name: Linear regression
+//description: Linear regression demo
```
Comment (skalkin) [inline packages/EDA/src/package.ts:551]: So is it a demo or not?

File: packages/EDA/src/package.ts [hunk 3]
```diff
@@ -542,3 +546,213 @@ export function anova(table: DG.DataFrame, factor: DG.Column, feature: DG.Column
...
+//name: Generate linear regression dataset
+//description: Create demo dataset for linear regression
+//input: int rowCount = 10000 {min: 1000; max: 10000000; step: 10000}
+//input: int colCount = 10 {min: 1; max: 1000; step: 10}
+//input: double featuresScale = 10 {min: -1000; max: 1000; step: 10}
+//input: double featuresBias = 10 {min: -1000; max: 1000; step: 10}
+//input: double paramsScale = 10 {min: -1000; max: 1000; step: 10}
+//input: double paramsBias = 10 {min: -1000; max: 1000; step: 10}
+//output: dataframe table
+export function generateDatasetForLinearRegressionTest(rowCount: number, colCount: number,
```
Comment (skalkin) [inline packages/EDA/src/package.ts:587]: We have a bunch of demo datasets in `grok.data.demo.xxx`, do we need another one here? If we really do - perhaps we need to move it to grok.data.demo or to utils somewhere to be reused?

File: packages/EDA/src/package.ts [hunk 4]
```diff
@@ -542,3 +546,213 @@ export function anova(table: DG.DataFrame, factor: DG.Column, feature: DG.Column
...
   runKNNImputer();
 }
+
+//name: Linear regression
+//description: Linear regression demo
+//input: dataframe table
+//input: column_list features {type: numerical}
+//input: column target {type: numerical}
+//input: bool plot = true {caption: plot}
+export async function linearRegression(table: DG.DataFrame, features: DG.ColumnList, target: DG.Column, plot: boolean): Promise<void> {
```
Comment (skalkin) [inline packages/EDA/src/package.ts:556]: Not sure what this function is doing - I mean, how is it making Datagrok better?
Does it modify the original table?

File: packages/EDA/src/package.ts [hunk 5]
```diff
@@ -542,3 +546,213 @@ export function anova(table: DG.DataFrame, factor: DG.Column, feature: DG.Column
...
+//meta.mlname: Softmax
+//meta.mlrole: isApplicable
+//input: dataframe df
+//input: string predict_column
+//output: bool result
+export function isApplicableSoftmax(df: DG.DataFrame, predict_column: string): boolean {
+  for (const col of df.columns) {
+    if (col.name !== predict_column) {
+      if ((col.type !== DG.COLUMN_TYPE.INT) && (col.type !== DG.COLUMN_TYPE.FLOAT) && (col.type !== DG.COLUMN_TYPE.QNUM) && (col.type !== DG.COLUMN_TYPE.BIG_INT))
+        return false;
```
Comment (skalkin) [inline packages/EDA/src/package.ts:697]: But what's wrong if a dataframe contains a string column? We are not forced to use it as a predictor.
isApplicable is only supposed to check whether a model could be applicable to that dataset (essentially if a dataset contains the right number of feature/predictor columns). Right, @KiK0S ?

File: packages/EDA/src/package.ts [hunk 6]
```diff
@@ -542,3 +546,213 @@ export function anova(table: DG.DataFrame, factor: DG.Column, feature: DG.Column
...
+  return (target.type === DG.COLUMN_TYPE.STRING);
+}
+
+//name: Softmax
+//desription: Softmax classifier demo
+//input: dataframe df
+//input: column_list features {type: numerical}
+//input: column target {type: string}
+//input: bool scatter = true
+export async function fitSoftmaxDemo(df: DG.DataFrame, features: DG.ColumnList, target: DG.Column, scatter: boolean) {
```
Comment (skalkin) [inline packages/EDA/src/package.ts:718]: I think we should come up with a better way for demo code, this just looks like a lot of code for such a simple purpose. Essentially all we need is a dataset, a model, and parameters, right? Perhaps there should be a helper method or something like this.

File: packages/EDA/src/softmax-classifier.ts [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+        this.categories = new Array<string>(modelDf.rowCount);
+
+        for (let i = 0; i < c; ++i) {
+          const c = columns.byIndex(i);
+          this.categories[i] = c.name;
+
+          if (c.type !== DG.COLUMN_TYPE.FLOAT)
+            throw new Error('incorrect params columns type');
+
+          this.params[i] = c.getRawData() as Float32Array;
```
Comment (skalkin) [inline packages/EDA/src/softmax-classifier.ts:97]: Are we sure that the column is going to contains floats and not ints, for instance? Perhaps - I did not manage to trace it to the root

File: packages/EDA/src/softmax-classifier.ts [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+        // extract params & categories
+        this.params = new Array<Float32Array>(c);
+        this.categories = new Array<string>(modelDf.rowCount);
+
+        for (let i = 0; i < c; ++i) {
+          const c = columns.byIndex(i);
+          this.categories[i] = c.name;
+
+          if (c.type !== DG.COLUMN_TYPE.FLOAT)
+            throw new Error('incorrect params columns type');
```
Comment (skalkin) [inline packages/EDA/src/softmax-classifier.ts:95]: Start the message with the capital letter. And it's "column type" not "columns type".
And I think more information is better. Like "Incorrect input column type. Expected: FLOAT, passed: STRING".

File: packages/EDA/src/svm.ts [hunk 9]
```diff
@@ -494,23 +473,30 @@ function getUnpackedModel(packedModel: any): any {
...
-    res = res && (columns.byIndex(i).type == 'double' || columns.byIndex(i).type == 'int');
-  return res;
-}
\ No newline at end of file
+
+  if ((labels.type !== DG.COLUMN_TYPE.STRING) || (labels.categories.length > 2))
+    return false;
+
+  for (const col of columns) {
+    if ((col.type !== DG.COLUMN_TYPE.FLOAT) && (col.type !== DG.COLUMN_TYPE.INT))
```
Comment (skalkin) [inline packages/EDA/src/svm.ts:497]: QNUM and BIGINT are not handled. I've seen it before somewhere, too. Let's use column.isNumerical?

File: packages/EDA/wasm/EDAAPI.js [hunk 10]
```diff
@@ -129,3 +129,48 @@ export async function _trainAndAnalyzeLSSVMInWebWorker(gamma, kernel, kernelPara
...
+      resolve(getResult(EDA['fitLinearRegressionParamsWithDataNormalizing'], e.data));
+    }
+  });
+}
+
+export function _fitLinearRegressionParams(features, targets, paramsCount) {
+  return callWasm(EDA, 'fitLinearRegressionParams', [features, targets, paramsCount]);
+}
+
+export async function _fitLinearRegressionParamsInWebWorker(features, targets, paramsCount) {
```
Comment (skalkin) [inline packages/EDA/wasm/EDAAPI.js:151]: Oh nice, didn't even know we were doing it in the web worker! But linear regressions is very fast - it's up to you, but it's fine to do it in the main thread I think. I guess do it the way it makes sense to keep the code simple.

**Review by skalkin — APPROVED** [review]
Wow, that's a lot of code - impressive! Looks good to me, but please address my comments. Please also think if we can make it smaller (by reducing demo, or exported function)

File: packages/EDA/wasm/softmax.h [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+					Z(i, j) = exp(Z(i, j)) * classWeights[i];
+			
+			sums = Z.colwise().sum();
+
+            // Compute predicted probabilities
+			for (int i = 0; i < m; ++i)
+				Z.col(i) = Z.col(i) / sums(i);			
+
+            // Copmute loss
+            loss = 0;
```
Comment (KiK0S) [inline packages/EDA/wasm/softmax.h:123]: is everything okay with indentation here? was concerned about coomments as well but here is is a line of code that is out of its scope

Comment (KiK0S) [inline packages/EDA/wasm/softmax.h:123]: is everything okay with indentation here? was concerned about comments as well but here is is a line of code that is out of its scope

File: packages/EDA/src/try.ts
Comment (KiK0S) [inline packages/EDA/src/try.ts:1]: i believe this should be excluded

File: packages/EDA/src/svm.ts [hunk 12]
```diff
@@ -494,23 +473,30 @@ function getUnpackedModel(packedModel: any): any {
 export async function getPrediction(df: DG.DataFrame, packedModel: any): Promise<DG.DataFrame> {
   const model = getUnpackedModel(new Uint8Array(packedModel));
 
+  console.log(model);
```
Comment (KiK0S) [inline packages/EDA/src/svm.ts:476]: do we need it?

File: packages/EDA/src/package.ts (hunk 3 above)
Comment (vmakarichev) [reply packages/EDA/src/package.ts:587]: Removed this method

File: packages/EDA/src/package.ts (hunk 4 above)
Comment (vmakarichev) [reply packages/EDA/src/package.ts:556]: It's a function for demo performance linear regression. REMOVED!

File: packages/EDA/src/package.ts (hunk 2 above)
Comment (vmakarichev) [reply packages/EDA/src/package.ts:551]: Demo. Removed!

File: packages/EDA/src/package.ts (hunk 5 above)
Comment (vmakarichev) [reply packages/EDA/src/package.ts:697]: Here, due to our predictive modeling features, df is a dataframe of column selected as features

File: packages/EDA/src/package.ts (hunk 6 above)
Comment (vmakarichev) [reply packages/EDA/src/package.ts:718]: It's just a function for demo softmax performance. Removed!

File: packages/EDA/src/softmax-classifier.ts (hunk 8 above)
Comment (vmakarichev) [reply packages/EDA/src/softmax-classifier.ts:95]: Modified

File: packages/EDA/src/svm.ts (hunk 9 above)
Comment (vmakarichev) [reply packages/EDA/src/svm.ts:497]: Fixed

File: packages/EDA/wasm/EDAAPI.js (hunk 10 above)
Comment (vmakarichev) [reply packages/EDA/wasm/EDAAPI.js:151]: This file is generated automatically by our c++-to-wasm export script. For each exported c++-function, this script generates two versions: in web worker call and in the main thread. After this, we can choose the preferable approach.
Currently, in linear regression, we use the second one

File: packages/EDA/src/try.ts
Comment (vmakarichev) [reply packages/EDA/src/try.ts:1]: Removed!

File: packages/EDA/src/svm.ts (hunk 12 above)
Comment (vmakarichev) [reply packages/EDA/src/svm.ts:476]: Removed

File: packages/EDA/wasm/softmax.h (hunk 11 above)
Comment (vmakarichev) [reply packages/EDA/wasm/softmax.h:123]: can you clarify?
